File: Android_GUI/src/com/github/chenxiaolong/dualbootpatcher/dialogs/GenericSingleChoiceDialog.java
/*
 * Copyright (C) 2015-2016  Andrew Gunnerson <andrewgunnerson@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.github.chenxiaolong.dualbootpatcher.dialogs;

import android.app.Dialog;
import android.os.Bundle;
import android.support.annotation.ArrayRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.Fragment;
import android.view.View;

import com.afollestad.materialdialogs.DialogAction;
import com.afollestad.materialdialogs.MaterialDialog;
import com.afollestad.materialdialogs.MaterialDialog.ListCallbackSingleChoice;
import com.afollestad.materialdialogs.MaterialDialog.SingleButtonCallback;

import java.io.Serializable;

public class GenericSingleChoiceDialog extends DialogFragment implements ListCallbackSingleChoice, SingleButtonCallback {
    private static final String ARG_BUILDER = "builder";
    private static final String ARG_TARGET = "target";
    private static final String ARG_TAG = "tag";

    private static final String EXTRA_SELECTED_INDEX = "selected_index";
    private static final String EXTRA_SELECTED_TEXT = "selected_text";

    private Builder mBuilder;
    private DialogListenerTarget mTarget;
    private String mTag;

    private String mSelectedText;

    public interface GenericSingleChoiceDialogListener {
        void onConfirmSingleChoice(@Nullable String tag, int index, String text);
    }

    @Nullable
    private GenericSingleChoiceDialogListener getOwner() {
        switch (mTarget) {
        case ACTIVITY:
            return (GenericSingleChoiceDialogListener) getActivity();
        case FRAGMENT:
            return (GenericSingleChoiceDialogListener) getTargetFragment();
        case NONE:
        default:
            return null;
        }
    }

    @Override
    @NonNull
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        Bundle args = getArguments();
        mBuilder = (Builder) args.getSerializable(ARG_BUILDER);
        mTarget = (DialogListenerTarget) args.getSerializable(ARG_TARGET);
        mTag = args.getString(ARG_TAG);

        int selectedIndex = -1;

        if (savedInstanceState != null) {
            selectedIndex = savedInstanceState.getInt(EXTRA_SELECTED_INDEX, -1);
            mSelectedText = savedInstanceState.getString(EXTRA_SELECTED_TEXT);
        }

        MaterialDialog.Builder builder = new MaterialDialog.Builder(getActivity());

        if (mBuilder.mTitle != null) {
            builder.title(mBuilder.mTitle);
        } else if (mBuilder.mTitleResId != 0) {
            builder.title(mBuilder.mTitleResId);
        }

        if (mBuilder.mMessage != null) {
            builder.content(mBuilder.mMessage);
        } else if (mBuilder.mMessageResId != 0) {
            builder.content(mBuilder.mMessageResId);
        }

        if (mBuilder.mPositive != null) {
            builder.positiveText(mBuilder.mPositive);
        } else if (mBuilder.mPositiveResId != 0) {
            builder.positiveText(mBuilder.mPositiveResId);
        }

        if (mBuilder.mNegative != null) {
            builder.negativeText(mBuilder.mNegative);
        } else if (mBuilder.mNegativeResId != 0) {
            builder.negativeText(mBuilder.mNegativeResId);
        }

        if (mBuilder.mChoices != null) {
            builder.items((CharSequence[]) mBuilder.mChoices);
        } else if (mBuilder.mChoicesResId != 0) {
            builder.items(mBuilder.mChoicesResId);
        }

        builder.itemsCallbackSingleChoice(selectedIndex, this);
        builder.alwaysCallSingleChoiceCallback();

        builder.onPositive(this);

        MaterialDialog dialog = builder.build();

        // Nothing selected by default, so don't enable OK button
        dialog.getActionButton(DialogAction.POSITIVE).setEnabled(selectedIndex >= 0);

        setCancelable(false);
        dialog.setCanceledOnTouchOutside(false);

        return dialog;
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        MaterialDialog dialog = (MaterialDialog) getDialog();
        if (dialog != null) {
            outState.putInt(EXTRA_SELECTED_INDEX, dialog.getSelectedIndex());
        }
        outState.putString(EXTRA_SELECTED_TEXT, mSelectedText);
    }

    @Override
    public boolean onSelection(MaterialDialog dialog, View itemView, int which, CharSequence text) {
        dialog.getActionButton(DialogAction.POSITIVE).setEnabled(true);
        mSelectedText = text.toString();
        return true;
    }

    @Override
    public void onClick(@NonNull MaterialDialog dialog, @NonNull DialogAction which) {
        GenericSingleChoiceDialogListener owner = getOwner();
        if (owner != null) {
            int index = dialog.getSelectedIndex();
            owner.onConfirmSingleChoice(mTag, index, mSelectedText);
        }
    }

    public static class Builder implements Serializable {
        @Nullable
        String mTitle;
        @StringRes
        int mTitleResId;
        @Nullable
        String mMessage;
        @StringRes
        int mMessageResId;
        @Nullable
        String mPositive;
        @StringRes
        int mPositiveResId;
        @Nullable
        String mNegative;
        @StringRes
        int mNegativeResId;
        @Nullable
        String[] mChoices;
        @ArrayRes
        int mChoicesResId;

        @NonNull
        public Builder title(@Nullable String title) {
            mTitle = title;
            mTitleResId = 0;
            return this;
        }

        @NonNull
        public Builder title(@StringRes int titleResId) {
            mTitle = null;
            mTitleResId = titleResId;
            return this;
        }

        @NonNull
        public Builder message(@Nullable String message) {
            mMessage = message;
            mMessageResId = 0;
            return this;
        }

        @NonNull
        public Builder message(@StringRes int messageResId) {
            mMessage = null;
            mMessageResId = messageResId;
            return this;
        }

        @NonNull
        public Builder positive(@Nullable String text) {
            mPositive = text;
            mPositiveResId = 0;
            return this;
        }

        @NonNull
        public Builder positive(@StringRes int textResId) {
            mPositive = null;
            mPositiveResId = textResId;
            return this;
        }

        @NonNull
        public Builder negative(@Nullable String text) {
            mNegative = text;
            mNegativeResId = 0;
            return this;
        }

        @NonNull
        public Builder negative(@StringRes int textResId) {
            mNegative = null;
            mNegativeResId = textResId;
            return this;
        }

        @NonNull
        public Builder choices(@Nullable String... choices) {
            mChoices = choices;
            mChoicesResId = 0;
            return this;
        }

        @NonNull
        public Builder choices(@ArrayRes int choicesResId) {
            mChoices = null;
            mChoicesResId = choicesResId;
            return this;
        }

        @NonNull
        public GenericSingleChoiceDialog buildFromFragment(@Nullable String tag,
                                                           @NonNull Fragment parent) {
            Bundle args = new Bundle();
            args.putSerializable(ARG_BUILDER, this);
            args.putSerializable(ARG_TARGET, DialogListenerTarget.FRAGMENT);
            args.putString(ARG_TAG, tag);

            GenericSingleChoiceDialog dialog = new GenericSingleChoiceDialog();
            dialog.setTargetFragment(parent, 0);
            dialog.setArguments(args);
            return dialog;
        }

        @NonNull
        public GenericSingleChoiceDialog buildFromActivity(@Nullable String tag) {
            Bundle args = new Bundle();
            args.putSerializable(ARG_BUILDER, this);
            args.putSerializable(ARG_TARGET, DialogListenerTarget.ACTIVITY);
            args.putString(ARG_TAG, tag);

            GenericSingleChoiceDialog dialog = new GenericSingleChoiceDialog();
            dialog.setArguments(args);
            return dialog;
        }
    }
}

File: Android_GUI/src/main/java/com/github/chenxiaolong/dualbootpatcher/dialogs/GenericSingleChoiceDialog.kt
/*
 * Copyright (C) 2015-2019  Andrew Gunnerson <andrewgunnerson@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.github.chenxiaolong.dualbootpatcher.dialogs

import android.app.Dialog
import android.os.Bundle
import androidx.annotation.ArrayRes
import androidx.annotation.StringRes
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.Fragment
import com.afollestad.materialdialogs.MaterialDialog
import com.afollestad.materialdialogs.WhichButton
import com.afollestad.materialdialogs.actions.setActionButtonEnabled
import com.afollestad.materialdialogs.list.listItemsSingleChoice
import java.io.Serializable

class GenericSingleChoiceDialog : DialogFragment() {
    private lateinit var target: DialogListenerTarget
    private lateinit var dialogTag: String

    private var selectedIndex: Int = -1
    private var selectedText: String? = null

    private val owner: GenericSingleChoiceDialogListener?
        get() {
            return when (target) {
                DialogListenerTarget.ACTIVITY -> activity as GenericSingleChoiceDialogListener?
                DialogListenerTarget.FRAGMENT -> targetFragment as GenericSingleChoiceDialogListener?
                DialogListenerTarget.NONE -> null
            }
        }

    interface GenericSingleChoiceDialogListener {
        fun onConfirmSingleChoice(tag: String, index: Int, text: String?)
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val args = arguments
        val builder = args!!.getSerializable(ARG_BUILDER) as Builder
        target = args.getSerializable(ARG_TARGET) as DialogListenerTarget
        dialogTag = args.getString(ARG_TAG)!!

        var selectedIndex = -1

        if (savedInstanceState != null) {
            selectedIndex = savedInstanceState.getInt(EXTRA_SELECTED_INDEX, -1)
            selectedText = savedInstanceState.getString(EXTRA_SELECTED_TEXT)
        }

        val dialog = MaterialDialog(requireActivity())

        if (builder.titleResId != null || builder.title != null) {
            dialog.title(builder.titleResId, text = builder.title)
        }

        if (builder.messageResId != null || builder.message != null) {
            dialog.message(builder.messageResId, text = builder.message)
        }

        if (builder.positiveResId != null || builder.positive != null) {
            dialog.positiveButton(builder.positiveResId, text = builder.positive) {
                owner?.onConfirmSingleChoice(dialogTag, selectedIndex, selectedText)
            }
        }

        if (builder.negativeResId != null || builder.negative != null) {
            dialog.negativeButton(builder.negativeResId, text = builder.negative)
        }

        if (builder.choicesResId != null || builder.choices != null) {
            @Suppress("CheckResult")
            dialog.listItemsSingleChoice(builder.choicesResId, items = builder.choices,
                    waitForPositiveButton = false) { _, index, text ->
                dialog.setActionButtonEnabled(WhichButton.POSITIVE, true)
                selectedIndex = index
                selectedText = text
            }
        }

        // Nothing selected by default, so don't enable OK button
        dialog.setActionButtonEnabled(WhichButton.POSITIVE, selectedIndex >= 0)

        isCancelable = false
        dialog.setCanceledOnTouchOutside(false)

        return dialog
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putInt(EXTRA_SELECTED_INDEX, selectedIndex)
        outState.putString(EXTRA_SELECTED_TEXT, selectedText)
    }

    class Builder : Serializable {
        internal var title: String? = null
        @StringRes
        internal var titleResId: Int? = null
        internal var message: String? = null
        @StringRes
        internal var messageResId: Int? = null
        internal var positive: String? = null
        @StringRes
        internal var positiveResId: Int? = null
        internal var negative: String? = null
        @StringRes
        internal var negativeResId: Int? = null
        internal var choices: List<String>? = null
        @ArrayRes
        internal var choicesResId: Int? = null

        fun title(title: String?): Builder {
            this.title = title
            titleResId = null
            return this
        }

        fun title(@StringRes titleResId: Int): Builder {
            title = null
            this.titleResId = titleResId
            return this
        }

        fun message(message: String?): Builder {
            this.message = message
            messageResId = null
            return this
        }

        fun message(@StringRes messageResId: Int): Builder {
            message = null
            this.messageResId = messageResId
            return this
        }

        fun positive(text: String?): Builder {
            positive = text
            positiveResId = null
            return this
        }

        fun positive(@StringRes textResId: Int): Builder {
            positive = null
            positiveResId = textResId
            return this
        }

        fun negative(text: String?): Builder {
            negative = text
            negativeResId = null
            return this
        }

        fun negative(@StringRes textResId: Int): Builder {
            negative = null
            negativeResId = textResId
            return this
        }

        fun choices(vararg choices: String): Builder {
            this.choices = listOf(*choices)
            choicesResId = null
            return this
        }

        fun choices(@ArrayRes choicesResId: Int): Builder {
            choices = null
            this.choicesResId = choicesResId
            return this
        }

        fun buildFromFragment(tag: String, parent: Fragment): GenericSingleChoiceDialog {
            val args = Bundle()
            args.putSerializable(ARG_BUILDER, this)
            args.putSerializable(ARG_TARGET, DialogListenerTarget.FRAGMENT)
            args.putString(ARG_TAG, tag)

            val dialog = GenericSingleChoiceDialog()
            dialog.setTargetFragment(parent, 0)
            dialog.arguments = args
            return dialog
        }

        fun buildFromActivity(tag: String): GenericSingleChoiceDialog {
            val args = Bundle()
            args.putSerializable(ARG_BUILDER, this)
            args.putSerializable(ARG_TARGET, DialogListenerTarget.ACTIVITY)
            args.putString(ARG_TAG, tag)

            val dialog = GenericSingleChoiceDialog()
            dialog.arguments = args
            return dialog
        }
    }

    companion object {
        private const val ARG_BUILDER = "builder"
        private const val ARG_TARGET = "target"
        private const val ARG_TAG = "tag"

        private const val EXTRA_SELECTED_INDEX = "selected_index"
        private const val EXTRA_SELECTED_TEXT = "selected_text"
    }
}
